File: flutroid_cli/lib/src/env.dart
import 'dart:io';

import 'package:path/path.dart' as p;

/// Reads settings from a `.env` file so the token is not a flag you retype.
///
/// The file is found by walking up from the app directory, so one `.env` at the
/// repo root covers every package under it. The server reads the same file.
///
/// Highest priority wins: a flag, then a real environment variable, then the
/// file. That order is what lets CI set a token with no `.env` checked out.

/// Parses `KEY=value` lines, allowing `export`, `#` comments and quotes.
Map<String, String> parseDotEnv(String source) {
  final values = <String, String>{};
  for (final raw in source.split('\n')) {
    var line = raw.trim();
    if (line.isEmpty || line.startsWith('#')) continue;
    if (line.startsWith('export ')) line = line.substring(7).trim();

    final eq = line.indexOf('=');
    if (eq <= 0) continue;

    final key = line.substring(0, eq).trim();
    var value = line.substring(eq + 1).trim();

    // Drop matching quotes, or trim a trailing comment off a bare value.
    if (value.length >= 2 &&
        ((value.startsWith('"') && value.endsWith('"')) ||
            (value.startsWith("'") && value.endsWith("'")))) {
      value = value.substring(1, value.length - 1);
    } else {
      final hash = value.indexOf(' #');
      if (hash >= 0) value = value.substring(0, hash).trim();
    }

    if (key.isNotEmpty) values[key] = value;
  }
  return values;
}

/// Finds the nearest `.env` at or above [start], or null.
File? findDotEnv(String start) {
  var dir = Directory(p.normalize(p.absolute(start)));
  while (true) {
    final candidate = File(p.join(dir.path, '.env'));
    if (candidate.existsSync()) return candidate;

    final parent = dir.parent;
    if (parent.path == dir.path) return null; // hit the filesystem root
    dir = parent;
  }
}

/// Loads the nearest `.env`, or an empty map when there is none.
Map<String, String> loadDotEnv({required String appDir}) {
  final file = findDotEnv(appDir) ?? findDotEnv(Directory.current.path);
  if (file == null) return const {};
  try {
    return parseDotEnv(file.readAsStringSync());
  } on IOException catch (error) {
    stderr.writeln('Warning: could not read ${file.path}: $error');
    return const {};
  }
}

/// Finds the upload token: `--token`, then the environment, then `.env`.
///
/// `UPLOAD_TOKEN` is accepted as well as `FLUTROID_TOKEN` because that is the
/// name the server uses, so one line in `.env` configures both.
String? resolveToken(String? explicit, {required String appDir}) {
  if (explicit != null && explicit.isNotEmpty) return explicit;

  for (final key in const ['FLUTROID_TOKEN', 'UPLOAD_TOKEN']) {
    final value = Platform.environment[key];
    if (value != null && value.isNotEmpty) return value;
  }

  final env = loadDotEnv(appDir: appDir);
  for (final key in const ['FLUTROID_TOKEN', 'UPLOAD_TOKEN']) {
    final value = env[key];
    if (value != null && value.isNotEmpty) return value;
  }
  return null;
}

/// Finds the backend URL, so `--server` can be left off too.
///
/// [explicitly] tells a typed `--server` apart from the default the parser
/// filled in; a typed one always wins.
String resolveServer(
  String fallback, {
  required String appDir,
  required bool explicitly,
}) {
  if (explicitly) return fallback;

  final fromEnv = Platform.environment['FLUTROID_SERVER'];
  if (fromEnv != null && fromEnv.isNotEmpty) return fromEnv;

  final value = loadDotEnv(appDir: appDir)['FLUTROID_SERVER'];
  return (value != null && value.isNotEmpty) ? value : fallback;
}

File: flutter_ota_server/src/env.ts
import fs from "node:fs";
import path from "node:path";

/// Loads the nearest `.env`, so the server and the CLI share one file at the
/// repo root.
///
/// Real environment variables win; the file only fills in what the process was
/// not already given. That is what lets a container set UPLOAD_TOKEN with no
/// `.env` on disk.

/** Parses `KEY=value` lines, allowing `export`, `#` comments and quotes. */
export function parseDotEnv(source: string): Record<string, string> {
  const values: Record<string, string> = {};
  for (const raw of source.split("\n")) {
    let line = raw.trim();
    if (!line || line.startsWith("#")) continue;
    if (line.startsWith("export ")) line = line.slice(7).trim();

    const eq = line.indexOf("=");
    if (eq <= 0) continue;

    const key = line.slice(0, eq).trim();
    let value = line.slice(eq + 1).trim();

    const quoted =
      value.length >= 2 &&
      ((value.startsWith('"') && value.endsWith('"')) ||
        (value.startsWith("'") && value.endsWith("'")));
    if (quoted) {
      value = value.slice(1, -1);
    } else {
      const hash = value.indexOf(" #");
      if (hash >= 0) value = value.slice(0, hash).trim();
    }

    if (key) values[key] = value;
  }
  return values;
}

/** Finds the nearest `.env` at or above [start]. */
export function findDotEnv(start: string): string | null {
  let dir = path.resolve(start);
  for (;;) {
    const candidate = path.join(dir, ".env");
    if (fs.existsSync(candidate)) return candidate;

    const parent = path.dirname(dir);
    if (parent === dir) return null; // filesystem root
    dir = parent;
  }
}

/**
 * Merges the nearest `.env` into `process.env`, never overwriting what is
 * already set. Returns the path it read, for the startup log.
 */
export function loadDotEnv(start = process.cwd()): string | null {
  const file = findDotEnv(start);
  if (!file) return null;

  try {
    for (const [key, value] of Object.entries(parseDotEnv(fs.readFileSync(file, "utf8")))) {
      if (process.env[key] === undefined) process.env[key] = value;
    }
    return file;
  } catch (error) {
    console.warn(`Could not read ${file}:`, error);
    return null;
  }
}
